### PR TITLE
Maintenance updates

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Update app according to [DataEng Application Maintenance document](https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/3445948428/DataEng+Application+Maintenance).

The following error codes were ignored: 
* tim/cli.py: ruff: noqa: TRY003, EM101`
   * [TRY003](https://docs.astral.sh/ruff/rules/raise-vanilla-args/): Opted to ignore this error except for cases when the `message` value exceeds line-length/is really long. 
   * [EM101](https://docs.astral.sh/ruff/rules/raw-string-in-exception/): Same reason as above

A few other things worth noting:
* There are some functions where a parameter is passed in because it is required by the `Click` command but isn't called. For these cases, the error code [ARG001](https://docs.astral.sh/ruff/rules/unused-function-argument/) had to be ignored
* Ignored missing imports for `ijson` and `smart_open` in [pyproject.toml](https://github.com/MITLibraries/timdex-index-manager/blob/maintenance-updates/pyproject.toml#L9-L11).
* The commits are essentially grouped according to the steps outlined in our maintenance docs, with pre-commit hooks being added at the end to ease development. 

### How can a reviewer manually see the effects of these changes?

1. Follow the instructions in [README | Index records into local OpenSearch Docker Container](https://github.com/MITLibraries/timdex-index-manager/blob/maintenance-updates/README.md#index-records-into-local-opensearch-docker-container), replacing:
* `<source-name>` -> `alma`
* `<filepath-to-records>` -> `tests/fixtures/sample_records.json`

2. After running the bulk-index in Step 4, you should see the following output:
```
2024-02-08 16:08:15,914 INFO tim.cli.bulk_index(): Bulk indexing records from file 'tests/fixtures/sample_records.json' into index 'alma-2024-02-08t21-05-56'
2024-02-08 16:08:16,084 INFO tim.opensearch.bulk_index(): All records ingested, refreshing index.
2024-02-08 16:08:16,143 INFO tim.cli.bulk_index(): Bulk indexing complete!
   Errors: 0
  Created: 6
  Updated: 0
  --------
    Total: 6
2024-02-08 16:08:16,143 INFO tim.cli.log_process_time(): Total time to complete process: 0:00:00.253575
```

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
NO

### What are the relevant tickets?
https://mitlibraries.atlassian.net/browse/IN-894

### Developer
- [x] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [x] All related Jira tickets are linked in commit message(s)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Review by ehanson8 — COMMENTED** [review]
A few comments but looking good!

File: setup.cfg
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply setup.cfg:1]: Ah, yes! Good catch. I kept telling myself to remember but alas. I'll take care of it!

File: README.md [hunk 1]
```diff
@@ -1,59 +1,158 @@
...
-DevTools is useful for writing/testing OpenSearch queries.
+3. Access OpenSearch Dashboards through <http://localhost:5601>.
 
-Discover is useful for browsing data. An index pattern will be required to use this tool. Note: do not set a date filed (choose the option to skip selecting a date field). It detects a date field in our indexes but then crashes trying to use it. Once you skip the data select field, just enter an index or alias to pull patterns from and it will automatically be configured to work well enough for initial data exploration.
+For a more detailed example with test data, please refer to the Confluence document: [How to run and query OpenSearch locally](https://mitlibraries.atlassian.net/wiki/spaces/D/pages/3586129972/How+to+run+and+query+OpenSearch+locally).
 
-### OpenSearch on AWS
+### Indexing a record into OpenSearch locally with Docker
+
+1. Follow the instructions in either[Running Opensearch locally with Docker](#running-opensearch-locally-with-docker) or [Running Opensearch and OpenSearch Dashboards locally with Docker](#running-opensearch-and-opensearch-dashboards-locally-with-docker).
```
Comment (ghukill) [inline README.md:59]: Tiny typo: need space between "either" and "Running"

File: README.md [hunk 2]
```diff
@@ -1,59 +1,158 @@
...
+2. To confirm the instance is up, run `pipenv run tim -u localhost ping` or visit http://localhost:9200/.
 
-DevTools is useful for writing/testing OpenSearch queries.
+3. Access OpenSearch Dashboards through <http://localhost:5601>.
 
-Discover is useful for browsing data. An index pattern will be required to use this tool. Note: do not set a date filed (choose the option to skip selecting a date field). It detects a date field in our indexes but then crashes trying to use it. Once you skip the data select field, just enter an index or alias to pull patterns from and it will automatically be configured to work well enough for initial data exploration.
+For a more detailed example with test data, please refer to the Confluence document: [How to run and query OpenSearch locally](https://mitlibraries.atlassian.net/wiki/spaces/D/pages/3586129972/How+to+run+and+query+OpenSearch+locally).
 
-### OpenSearch on AWS
+### Indexing a record into OpenSearch locally with Docker
```
Comment (ghukill) [inline README.md:57]: I might suggest rewording to "Index sample records into local Opensearch Docker container".

Two reasons:

1. It's not a single record, but a batch of records in a single file
2. The code provided is running the native python code, but it will index _into_ a local Opensearch container.

File: Makefile [hunk 3]
```diff
@@ -6,67 +6,81 @@ ECR_URL_DEV:=222053980223.dkr.ecr.us-east-1.amazonaws.com/timdex-index-manager-d
...
 
-### Dependency commands ###
+help: # preview Makefile commands
+	@awk 'BEGIN { FS = ":.*#"; print "Usage:  make <target>\n\nTargets:" } \
+/^[-_[:alpha:]]+:.?*#/ { printf "  %-15s%s\n", $$1, $$2 }' $(MAKEFILE_LIST)
 
-install: ## Install dependencies and CLI app
+## Dependency commands 
+
+install: # install Python dependencies and pre-commit hook
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply Makefile:15]: See Step 3 in [DataEng Application Maintenance](https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/3445948428/DataEng+Application+Maintenance) doc.

File: README.md [hunk 4]
```diff
@@ -1,59 +1,158 @@
...
+│ delete       Delete an index.                                                                                                       │
+│ promote      Promote index as the primary alias and add it to any additional provided aliases.                                      │
+│ demote       Demote an index from all its associated aliases.                                                                       │
+╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
+╭─ Bulk record processing commands ───────────────────────────────────────────────────────────────────────────────────────────────────╮
+│ bulk-index                         Bulk index records into an index.                                                                │
+│ bulk-delete                        Bulk delete records from an index.                                                               │
+╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```
Comment (ghukill) [inline README.md:156]: Perhaps time will tell if this pattern is a good one... as we do need to keep this up to date with changes, but I must admit I like seeing it in the readme.  This allows seeing the CLI help without cloning, installing, and setting anything up, which is handy.

I spoke with Jeremy awhile back about potentially automating this, which I think could be pretty neat.  Perhaps it would write to a standalone file, something like `docs/cli_help.md`?  But, for the time being, I'm appreciating this in the base README.

File: setup.cfg
[the remark this answers is not in the document]
Comment (ghukill) [reply setup.cfg:1]: Good catch, I think I've missed that too.

File: tests/conftest.py [hunk 5]
```diff
@@ -1,35 +1,35 @@
...
 import vcr
 from click.testing import CliRunner
 
 from tim.opensearch import configure_opensearch_client
 
+EXIT_CODES = {
+    "success": 0,
+    "error": 1,
+    "invalid_command": 2,
+}
```
Comment (ghukill) [inline tests/conftest.py:11]: Seconding @ehanson8 here, nice pattern.

File: tests/conftest.py [hunk 6]
```diff
@@ -1,35 +1,35 @@
...
-
-
-@pytest.fixture()
+def _test_env(monkeypatch):
+    monkeypatch.setenv("SENTRY_DSN", "None")
+    monkeypatch.setenv("WORKSPACE", "test")
+    monkeypatch.setenv("AWS_ACCESS_KEY_ID", "test")
+    monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "test")
+    monkeypatch.setenv("AWS_SESSION_TOKEN", "test")
+    monkeypatch.setenv("TIMDEX_OPENSEARCH_ENDPOINT", "localhost")
```
Comment (ghukill) [inline tests/conftest.py:25]: Good call on the `monkeypatch` update; this has proven much easier to work with in other repos.

File: tests/test_helpers.py [hunk 7]
```diff
@@ -69,53 +68,56 @@ def test_get_source_from_index_without_dash():
 
 def test_parse_records():
     records = list(helpers.parse_records("tests/fixtures/sample_records.json"))
-    assert len(records) == 6
+    n_sample_records = 6
+    assert len(records) == n_sample_records
```
Comment (ghukill) [inline tests/test_helpers.py:72]: Not requesting a change here, but in my opinion this "magic comparison" thing linting error is probably not something we need to worry about in tests; I think it's okay to use direct number comparisons for simple logic like this.  Appreciate the effort to meet high linting requirements, but going forward I'd propose it's not needed.

File: tests/test_helpers.py [hunk 8]
```diff
@@ -69,53 +68,56 @@ def test_get_source_from_index_without_dash():
...
 
 @my_vcr.use_cassette("helpers/bulk_cli_no_primary_index_for_source.yaml")
 def test_validate_bulk_cli_options_no_primary_index_for_source(test_opensearch_client):
-    with pytest.raises(BadParameter) as error:
+    with pytest.raises(
+        BadParameter,
+        match=(
+            "No index name was passed and there is no "
+            "primary-aliased index for source 'dspace'."
+        ),
```
Comment (ghukill) [inline tests/test_helpers.py:118]: +1 to including `match` when possible; helps to understand the exception thrown without digging into that code.

**Review by ghukill — CHANGES_REQUESTED** [review]
Submitting a "Request Changes" for a small typo in the README, but otherwise looks great to me!  Thanks for the thoughtful maintenance work here.

With that typo changed, please consider my review "Approved".

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (ehanson8) [reply Makefile:15]: Excellent, thanks!

File: setup.cfg
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply setup.cfg:1]: Deleted!

File: README.md (hunk 4 above)
Comment (jonavellecuerdo) [reply README.md:156]: Automating it sounds epic!

File: tests/test_helpers.py (hunk 7 above)
Comment (jonavellecuerdo) [reply tests/test_helpers.py:72]: I'll keep this in mind!

File: README.md (hunk 2 above)
Comment (jonavellecuerdo) [reply README.md]: I updated the heading to read "Index records into local OpenSearch Docker container". Left out the world "sample" to keep it generalizable 🤓 .

**Review by ehanson8 — APPROVED** [review]

**Review by ghukill — APPROVED** [review]
